lib/mix/test/mix/tasks/do_test.exs

```diff
@@ -24,4 +24,115 @@ defmodule Mix.Tasks.DoTest do
   end
+
+  test "runs given tasks for a single app specified by app flag" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      Mix.Project.in_project(:umbrella, ".", fn _ ->
+        Mix.Tasks.Do.run(["--app", "bar", "compile", "--list,", "cmd", "echo", "hello"])
+
+        nl = os_newline()
+        assert_received {:mix_shell, :info, ["==> bar"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+        refute_received {:mix_shell, :info, ["==> foo"]}
+        refute_received {:mix_shell, :run, ["hello" <> ^nl]}
+      end)
+    end)
+  end
+
+  test "runs given tasks for each app specified by app flag" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      Mix.Project.in_project(:umbrella, ".", fn _ ->
+        Mix.Tasks.Do.run([
+          "--app",
+          "bar",
+          "--app",
+          "foo",
+          "compile",
+          "--list,",
+          "cmd",
+          "echo",
+          "hello"
+        ])
+
+        nl = os_newline()
+        assert_received {:mix_shell, :info, ["==> bar"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+        assert_received {:mix_shell, :info, ["==> foo"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+      end)
+    end)
+  end
+
+  test "runs non-recursive tasks at project root level" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      Mix.Project.in_project(:umbrella, ".", fn _ ->
+        Mix.Tasks.Do.run(["--app", "bar", "--app", "foo", "compile", "--list,", "help"])
+
+        assert_received {:mix_shell, :info, ["mix help" <> _]}
+        assert_received {:mix_shell, :info, ["mix compile.app" <> _]}
+      end)
+    end)
+  end
+
+  test "raises when -app is given but the project is not an umbrella" do
+    assert_raise Mix.Error,
+                 "Could not run \"cmd\" with the --app option because this is not an umbrella project",
+                 fn ->
+                   Mix.Tasks.Do.run(["--app", "bar", "--app", "foo", "cmd", "echo", "hello"])
+                 end
+  end
+
+  test "runs given aliases for each app specified by app flag" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      aliases = [
+        e: ["cmd echo hello"],
+        p: fn val -> Mix.shell().info(inspect(val)) end
+      ]
+
+      Mix.Project.in_project(:umbrella, ".", [aliases: aliases], fn _ ->
+        Mix.Tasks.Do.run(["compile"])
+        Mix.Tasks.Do.run(["--app", "bar", "--app", "foo", "e,", "p", "Foo"])
+
+        nl = os_newline()
+        assert_received {:mix_shell, :info, ["==> bar"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+        assert_received {:mix_shell, :info, ["[\"Foo\"]"]}
+
+        assert_received {:mix_shell, :info, ["==> foo"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+        assert_received {:mix_shell, :info, ["[\"Foo\"]"]}
+      end)
+    end)
+  end
+
+  test "runs non-recursive aliases at project root level" do
```

Comment:
@josevalim Do you still think this edge case is well resolved? I initially emitted a warning for this kind of tasks (because the -app argument is basically ignored here) but we reverted because of this https://github.com/elixir-lang/elixir/pull/11593#pullrequestreview-862038204

Since we are now raising an error when --app is given in no-umbrella projects, I was thinking again in this situation and I don't like the fact we are still running the task (at the root level) and ignoring the --app option, without any kind of warning/error.

Maybe you're still OK, I'm just mentioning it just in case you want to double check :)

Reply:
For now I think this is the expected behaviour. If a task is recursive or not is kind of an implementation detail. We will see what comes out of this in practice. :)